File: tests/xcalablemp/NOT-IMPLEMENTED/59/reduction_async14.c
/*testp053.c*/
/*reduction指示文(&)のテスト*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>
#include<stdlib.h>   
#pragma xmp nodes p(4,*)
int procs,id;
int mask,val1,val2;
char *result;
int i,w;
int main(void){

   if(xmp_num_nodes() > 31){
      printf("%s\n","You have to run this program by less than 32 nodes.");
   }

   procs = xmp_num_nodes();
   id = xmp_num_nodes()-1;
   result = "OK";
   w=1;
   for(i=0;i<procs;i++){
      w*2;
   }
   for(i=0;i<w;i=i+2){
      mask = 1 << id;
      val1 =!(i & mask);
#pragma xmp reduction(&:val1) async(1)
      mask = 1 << id;
      val2 =!((i+1) & mask);
#pragma xmp reduction(&:val2) async(2)
#pragma xmp wait_async(1)
      if((!(val1)) != i){
         result = "NG";
      }
#pragma xmp wait_async(2)
      if(!val2 != i+1){
         result = "NG";
      }

   }

   printf("%d %s %s\n",xmp_node_num(),"testp053.c",result); 
   return 0;
}
      
         
      
   

File: tests/xcalablemp/NOT-IMPLEMENTED/59/reduction_async14.f90
! testp053.f
! reduction指示文(iand)のテスト

      program main
      include 'xmp_lib.h'
!$xmp nodes p(4,*)
      integer procs, id
      integer mask, val1, val2
      character(len=2) result

      if(xmp_num_nodes().gt.31) then
         print *, 'You have to run this program by less than 32 nodes.'
      endif


      procs = xmp_num_nodes()
      id = xmp_node_num()-1
      result = 'OK'
      do i=0, 2**procs-1, 2
         mask = lshift(1, id)
         val1 = not(iand(i, mask))
!$xmp reduction(iand: val1) async(1)
         mask = lshift(1, id)
         val2 = not(iand(i+1, mask))
!$xmp reduction(iand: val2) async(2)
!$xmp wait_async(1)
         if(not(val1) .ne. i) then
            result = 'NG'
         endif
!$xmp wait_async(2)
         if(not(val2) .ne. i+1) then
            result = 'NG'
         endif
      enddo

      print *, xmp_node_num(), 'testp053.f ', result

      end
